design/bgp-bfd.md

```diff
@@ -0,0 +1,164 @@
+# BFD enablement in BGP
+
+## Summary
+
+The purpose of this enhancement is to use enable [BFD](https://en.wikipedia.org/wiki/Bidirectional_Forwarding_Detection) to be used in conjunction with BGP.
+
+BFD will be available only in the upcoming [FRR based implementation](0001-frr.md).
+
+## Motivation
+
+BFD provides a quicker path failure detection than BGP, and using them together will allow users to provide a better
+service.
+
+BFD for BGP is supported by FRR [out of the box](http://docs.frrouting.org/en/latest/bfd.html#bgp-bfd-configuration).
+
+### Goals
+
+* Provide a way to establish BFD on BGP session.
+* Provide a way to configure the parameters of the BFD session associated with a BGP session.
+
+### Non-Goals
+
+* Implement BFD on the legacy BGP implementation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to declare a BGP session to be backed up by BFD.
+
+#### Story 2
+
+As a cluster administrator, I want to be able to set the BFD parameters related to a BGP session.
+
+## Design Details
+
+### BFD Design proposal
+
+The idea is to leverage FRR to enable BFD on a BGP session.
+
+When declaring a BFD peer, all it takes to enable BFD is to add a `bfd` property on the neighbour:
+
+```yaml
+    neighbor <A.B.C.D|X:X::X:X|WORD> bfd profile BFDPROF
+```
+
+The proposal here is to define a bfd profile section in the config structure that looks like:
+
+```yaml
+    bfd-profiles:
+    - name: bfdprofile1
+      receive-interval: 150
+      transmit-interval: 150
+      detect-multiplier: 10
+      echo-receive-interval: 20
+      echo-transmit-interval: 20
+      echo-mode: true
+      passive-mode: true
+      minimum-ttl: 5
+```
+
+When a property of the profile is not set, MetalLB will honor [FRR default values](https://docs.frrouting.org/en/latest/bfd.html#peer-profile-configuration).
+
+When setting a BGP peer, an optional bfd-profile property will enable BFD:
+
+```yaml
+    peers:
+    - peer-address: 10.0.0.1
+      peer-asn: 64501
+      my-asn: 64500
+      bfd-profile: bfdprofile1
+```
+
+#### Impacts on the operator
+
+Keeping the operator as the guinea pig for the CRD implementation, a new BFDProfile CRD will be introduced, with the form of:
+
+```yaml
+apiVersion: metallb.io/v1alpha1
+kind: BFDProfile
+metadata:
+  name: profile
+  namespace: metallb-system
+spec:
+  receive-interval: 150
+  transmit-interval: 150
+  detect-multiplier: 10
+```
+
+Similarly, the BGPPeer CRD that is getting introduced will be configured with a new optional `bfdProfile` field.
+
+Ideally, if/when the CRDs are moved back to MetalLB, it will be possible to enrich the BGPPeer CRD status with information containing the status of the BFD session.
```

Comment:
What I meant is. ff the CRDs are handled by metallb itself, it will be possible to enrich the status with informations that metallb is aware of.
It is possible to publish that information (such as session state) as metrics, but it would be nice to do something like
`kubectl get peer foo` and see the status of the sessions at a glance. This is possible only if the CRDs are handled by metallb itself.

Reply:
I mean the operator could consume the metrics and then reflect status via the CRD ... obviously not ideal, but doesn't seem to be completely blocked on living it metallb

Reply:
lol, yes, you're right.  I wasn't arguing I had a good solution, I was just splitting hairs that it was technically possible!  Anyway, I'll resolve this thread.  :)